memstore/upsert_batch.go

```diff
@@ -520,97 +573,63 @@ func readUpsertBatchNew(buffer []byte) (*UpsertBatch, error) {
 
 }
 
-// TODO: delete after upsert batch new version migration
-func readUpsertBatchOld(buffer []byte) (*UpsertBatch, error) {
-	batch := &UpsertBatch{
-		buffer:      buffer,
-		columnsByID: make(map[int]int),
-	}
-
-	// numRows.
-	reader := utils.NewBufferReader(buffer)
-
-	numRows, err := reader.ReadInt32(0)
-	if err != nil {
-		return nil, utils.StackError(err, "Failed to read number of rows")
-	}
-	if numRows < 0 {
-		return nil, utils.StackError(err, "Number of rows %d should be >= 0", numRows)
-	}
-
-	batch.NumRows = int(numRows)
-	// numColumns.
-	numColumns, err := reader.ReadUint16(4)
-	if err != nil {
-		return nil, utils.StackError(err, "Failed to read number of columns")
-	}
-	batch.NumColumns = int(numColumns)
-
-	// Header too small, error out.
-	if len(buffer) < 8+common.ColumnHeaderSize(batch.NumColumns) {
-		return nil, utils.StackError(nil, "Invalid upsert batch data with incomplete header section")
-	}
-
-	header := common.NewUpsertBatchHeader(buffer[8:], batch.NumColumns)
-
-	columns := make([]*columnReader, batch.NumColumns)
-	for i := range columns {
-		columnType, err := header.ReadColumnType(i)
-		if err != nil {
-			return nil, utils.StackError(err, "Failed to read type for column %d", i)
-		}
+// ResolveEnumDict resolves upsert batch level enum dict with instance level enum dictionary
+func (u *UpsertBatch) ResolveEnumDict(table string, tableSchema *TableSchema, metaStore metastore.MetaStore) error {
+	tableSchema.RLock()
+	enumDicts := tableSchema.EnumDicts
+	columnDeletions := tableSchema.GetColumnDeletions()
+	tableSchema.RUnlock()
 
-		columnID, err := header.ReadColumnID(i)
-		if err != nil {
-			return nil, utils.StackError(err, "Failed to read id for column %d", i)
+	for _, columnReader := range u.columns {
+		if !memCom.IsEnumType(columnReader.dataType) {
+			continue
 		}
-		batch.columnsByID[columnID] = i
-
-		columnMode, columnUpdateMode, err := header.ReadColumnFlag(i)
-		if err != nil {
-			return nil, utils.StackError(err, "Failed to read mode for column %d", i)
+		columnID := columnReader.columnID
+		if columnID > len(columnDeletions) || columnDeletions[columnID] {
+			continue
 		}
-
-		columns[i] = &columnReader{columnID: columnID, columnMode: columnMode, columnUpdateMode: columnUpdateMode, dataType: columnType,
-			cmpFunc: common.GetCompareFunc(columnType)}
-
-		columnStartOffset, err := header.ReadColumnOffset(i)
-		if err != nil {
-			return nil, utils.StackError(err, "Failed to read start offset for column %d", i)
+		if tableSchema.Schema.Columns[columnID].DisableAutoExpand {
+			// no enum resolution needed, in this case, enum will be part of table schema
+			// and translation will be done by ingestion client or subscriber
```

Comment:
how can subscriber do the translation? it doesn't know about each instance's local enum dict, right?

Reply:
this is for columns that auto expand is disabled, in this situation, the enum expansion will need to be done by the user through schema change in controller, therefore, the enum list for those columns will be part of the table schema. In such case, subscriber or client will be able to translate the enum.